File: golang/ssh/auto-replicas/auto-replicas.go
/*
 * @Author: magician
 * @Date: 2024-08-27 23:33:24
 * @LastEditors: magician
 * @LastEditTime: 2024-09-11 00:28:43
 * @FilePath: /go/src/go_code/ssh/auto-replicas/auto-replicas.go
 * @Description:
 *
 * Copyright (c) 2024 by ${git_name_email}, All Rights Reserved.
 */
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"slices"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/Shopify/sarama"
	"github.com/pkg/sftp"
	"github.com/spf13/viper"
	"golang.org/x/crypto/ssh"
	"gopkg.in/yaml.v2"
)

func get_kafka_lag(brokerList []string, limit int64) (map[string]bool, error) {
	/**
	@brief 获取kafka groupid
	@param brokerList []string kafka服务器地址
	@param limit int64 lag大小
	@return map[string]bool groupid
	@return error
	**/
	config := sarama.NewConfig()
	config.Version = sarama.V2_1_0_0
	config.Consumer.Return.Errors = true

	client, err := sarama.NewClient(brokerList, config)
	if err != nil {
		return nil, err
	}
	defer client.Close()

	admin, err := sarama.NewClusterAdminFromClient(client)
	if err != nil {
		return nil, err
	}
	defer admin.Close()

	// 读取kafka所有的groupid
	// groupIds, err := admin.ListConsumerGroups()
	// if err != nil {
	// 	return nil, err
	// }

	// for groupId, _ := range groupIds {
	// 	fmt.Println(groupId)
	// }

	type Result struct {
		groupId string
		lag     map[string]int64
		err     error
	}

	wxgz_groups := []string{"groupid_threat", "groupid_dataservice", "event_data_preparation", "cleaners_groupid", "groupid_running_distribution"}
	// lags := make(map[string]map[string]int64)
	results := make(chan Result, len(wxgz_groups))
	var wg sync.WaitGroup

	for _, groupId := range wxgz_groups {
		wg.Add(1)
		go func(groupId string) {
			defer wg.Done()
			groupLag := make(map[string]int64)
			consecutiveLags := 0
			for i := 0; i < 3; i++ {

				groups, err := admin.ListConsumerGroupOffsets(groupId, nil)
				if err != nil {
					results <- Result{groupId, nil, err}
					return
				}
				totalLag := int64(0)
				for topic, partitions := range groups.Blocks {
					for partition, block := range partitions {
						newestOffset, err := client.GetOffset(topic, partition, sarama.OffsetNewest)
						if err != nil {
							results <- Result{groupId, nil, err}
							return
						}
						groupLag[topic] += newestOffset - block.Offset
					}
				}
				if totalLag >= limit {
					consecutiveLags++
					time.Sleep(5 * time.Minute)
					// time.Sleep(5 * time.Second)
				} else {
					consecutiveLags = 0
				}
			}
			if consecutiveLags >= 3 {
				results <- Result{groupId, groupLag, nil}
			}
		}(groupId)
	}

	wg.Wait()
	close(results)

	// 处理结果
	uniqueGroupIds := make(map[string]bool)
	for result := range results {
		if result.err != nil {
			fmt.Printf("Error processing group %s: %v\n", result.groupId, result.err)
		} else {
			uniqueGroupIds[result.groupId] = true
		}
	}

	return uniqueGroupIds, nil
}

func get_remote_system_info(ip, username, password string, port int) (string, string, string, error) {
	/**
	@brief 获取远程系统信息
	@param ip string 服务器ip
	@param username string 用户名
	@param password string 密码
	@param port int 端口号
	@return string CPU信息
	@return string 内存信息
	@return string 磁盘信息
	**/

	config := &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{
			ssh.Password(password),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         5 * time.Second,
	}

	// 连接到远程服务器
	hosts := ip + ":" + fmt.Sprintf("%d", port)
	client, err := ssh.Dial("tcp", hosts, config)
	if err != nil {
		fmt.Println("Failed to dial: ", err)
		return "", "", "", err
	}
	defer client.Close()

	// 执行远程命令
	session, err := client.NewSession()
	if err != nil {
		fmt.Println("Failed to create session: ", err)
		return "", "", "", err
	}
	defer session.Close()
	// 执行远程命令
	output, err := session.CombinedOutput("top -b -n 1 | grep 'Cpu(s)'")
	if err != nil {
		fmt.Println("Failed to run: ", err)
		return "", "", "", err
	}
	cpuinfo := strings.Split(string(output), ",")[3]
	cpuinfo = strings.Split(cpuinfo, "id")[0]
	// fmt.Printf("cpu可用为: %s\n", cpuinfo)

	// 执行远程命令
	session, err = client.NewSession()
	if err != nil {
		fmt.Println("Failed to create session: ", err)
		return "", "", "", err
	}
	defer session.Close()
	output, err = session.CombinedOutput("free -g|grep Mem")
	if err != nil {
		fmt.Println("Failed to run: ", err)
		return "", "", "", err
	}
	meminfo := strings.Fields(string(output))[6]
	// fmt.Printf("mem可用为: %s\n", meminfo)

	// 执行远程命令
	session, err = client.NewSession()
	if err != nil {
		fmt.Println("Failed to create session: ", err)
		return "", "", "", err
	}
	defer session.Close()
	output, err = session.CombinedOutput("df -h -BG /home/app | sort -k 4 -h|tail -n 1")
	if err != nil {
		fmt.Println("Failed to run: ", err)
		return "", "", "", err
	}
	disk_output := strings.Fields(string(output))[3]
	disk_output = strings.TrimSuffix(disk_output, "G")
	// fmt.Printf("硬盘可用为: %s\n", disk_output)
	return cpuinfo, meminfo, disk_output, nil

}

func get_app_replicas(ip, username, password string, port int, app_name string) (int, error) {
	/*
		@brief 获取应用副本数
		@param ip string 服务器ip
		@param username string 用户名
		@param password string 密码
		@param port int 端口号
		@param app_name string 应用名称
		@return int 应用副本数
		**/
	sshConfig := &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{
			ssh.Password(password),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         5 * time.Second,
	}
	addr := fmt.Sprintf("%s:%d", ip, port)
	conn, err := ssh.Dial("tcp", addr, sshConfig)
	if err != nil {

		log.Fatal("Failed to dial: ", err)
	}
	defer conn.Close()
	client, err := sftp.NewClient(conn)
	if err != nil {
		log.Fatal("Failed to create client: ", err)
	}
	defer client.Close()
	remotePath := filepath.Join("/home/app/app", app_name, "bin/docker-compose.yml")
	file, err := client.Open(remotePath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	content, err := ioutil.ReadAll(file)
	if err != nil {
		log.Fatal(err)
	}
	// Unmarshal YAML data
	var composeData map[string]interface{}
	if err := yaml.Unmarshal(content, &composeData); err != nil {
		log.Fatalf("Failed to unmarshal YAML: %s", err)
	}
	replicas := int(0)
	// Update the 'replicas' field for the specified service
	if services, ok := composeData["services"].(map[interface{}]interface{}); ok {
		for _, service := range services {
			if deploy, ok := service.(map[interface{}]interface{})["deploy"].(map[interface{}]interface{}); ok {
				replicas = deploy["replicas"].(int)

				// fmt.Println(deploy["replicas"])
				break
			}
		}
	}
	return replicas, nil
}

func set_app_replicas(ip, username, password string, port int, app_name string, replicas int) {
	/*
		@brief 修改应用副本数
		@param ip string 服务器ip
		@param username string 用户名
		@param password string 密码
		@param port int 端口号
		@param app_name string 应用名称
		@param replicas int 副本数
		**/
	sshConfig := &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{
			ssh.Password(password),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         5 * time.Second,
	}
	addr := fmt.Sprintf("%s:%d", ip, port)
	conn, err := ssh.Dial("tcp", addr, sshConfig)
	if err != nil {

		log.Fatal("Failed to dial: ", err)
	}
	defer conn.Close()
	client, err := sftp.NewClient(conn)
	if err != nil {
		log.Fatal("Failed to create client: ", err)
	}
	defer client.Close()
	remotePath := filepath.Join("/home/app/app", app_name, "bin/docker-compose.yml")
	file, err := client.Open(remotePath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	content, err := ioutil.ReadAll(file)
	if err != nil {
		log.Fatal(err)
	}
	// Unmarshal YAML data
	var composeData map[string]interface{}
	if err := yaml.Unmarshal(content, &composeData); err != nil {
		log.Fatalf("Failed to unmarshal YAML: %s", err)
	}
	// Update the 'replicas' field for the specified service
	if services, ok := composeData["services"].(map[interface{}]interface{}); ok {
		for _, service := range services {
			if deploy, ok := service.(map[interface{}]interface{})["deploy"].(map[interface{}]interface{}); ok {
				deploy["replicas"] = replicas
				// fmt.Println(deploy["replicas"])
				break
			}
		}
	}
	updateData, err := yaml.Marshal(composeData)
	if err != nil {
		log.Fatal("Failed to marshal YAML: ", err)
	}
	file, err = client.Create(remotePath)
	if err != nil {
		log.Fatal("Failed to write file: ", err)
	}
	defer file.Close()

	if _, err := file.Write(updateData); err != nil {
		log.Fatal("Failed to write file: ", err)
	}
}

func app_restart(ip, username, password string, port int, app_name string) {
	/*
		@brief 应用重启
		@param ip string 服务器ip
		@param username string 用户名
		@param password string 密码
		@param port int 端口号
		@param app_name string 应用名称
	**/

	// 创建ssh客户端
	sshConfig := &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{
			ssh.Password(password),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         30 * time.Second,
	}
	addr := fmt.Sprintf("%s:%d", ip, port)
	client, err := ssh.Dial("tcp", addr, sshConfig)
	if err != nil {
		log.Fatal("无法连接到 %s: %s", addr, err)
	}
	defer client.Close()
	// 执行远程命令
	session, err := client.NewSession()
	if err != nil {
		fmt.Println("Failed to create session: ", err)
		return
	}
	defer session.Close()
	_, err = session.CombinedOutput("/home/app/bin/docker-compose -f /home/app/app/" + app_name + "/bin/docker-compose.yml down")
	if err != nil {
		log.Fatal(err)
	}
	// fmt.Print(output)
	time.Sleep(2 * time.Second)
	// 执行远程命令
	session, err = client.NewSession()
	if err != nil {
		fmt.Println("Failed to create session: ", err)
		return
	}
	defer session.Close()
	_, err = session.CombinedOutput("/home/app/bin/docker-compose -f /home/app/app/" + app_name + "/bin/docker-compose.yml up -d")
	if err != nil {
		log.Printf("启动服务失败: %v", err)
	}
	// fmt.Print(output)
}

func deploy_app_replicas(v *viper.Viper, app_name string) {
	/**
		@brief 部署应用副本数
		@param app_name string 应用名称
	**/
	username := v.GetStringMapString("database.ssh")["username"]
	password := v.GetStringMapString("database.ssh")["password"]
	port_str := v.GetStringMapString("database.ssh")["port"]
	port, _ := strconv.Atoi(port_str)
	replicas := v.Get("database.replicas").(map[string]interface{})
	limit_cpu_str := v.GetStringMapString("database.limit")["cpu"]
	limit_cpu, _ := strconv.ParseFloat(limit_cpu_str, 32)
	limit_mem_str := v.GetStringMapString("database.limit")["memory"]
	limit_mem, _ := strconv.Atoi(limit_mem_str)
	limit_disk_str := v.GetStringMapString("database.limit")["disk"]
	limit_disk, _ := strconv.ParseFloat(limit_disk_str, 32)

	// 检查 appName 是否存在于 replicas 中
	if service, exists := replicas[app_name]; exists {
		// 将 service 转换为切片
		for _, serviceConfig := range service.([]interface{}) {
			config := serviceConfig.(map[string]interface{})
			if config["enable"].(bool) {
				ip := config["ip"].(string)
				// min := config["min"].(int)
				max := config["max"].(int)
				cpuinfo_str, meminfo_str, disk_output_str, _ := get_remote_system_info(ip, username, password, port)
				cpuinfo_str = strings.TrimSpace(cpuinfo_str)
				cpuinfo, _ := strconv.ParseFloat(cpuinfo_str, 32)
				meminfo, _ := strconv.Atoi(meminfo_str)
				disk_output, _ := strconv.ParseFloat(disk_output_str, 32)
				if cpuinfo >= limit_cpu && meminfo >= limit_mem && disk_output >= limit_disk {
					fmt.Printf("cpuinfo: %v\nmeminfo: %v\ndisk_output: %v\n", cpuinfo, meminfo, disk_output)
					replicas, _ := get_app_replicas(ip, username, password, port, app_name)
					fmt.Printf("[%s]扩容前: 当前副本数:%d\n", app_name, replicas)
					if replicas < max {
						set_app_replicas(ip, username, password, port, app_name, replicas+1)
						app_restart(ip, username, password, port, app_name)
						replicas, _ = get_app_replicas(ip, username, password, port, app_name)
						fmt.Printf("[%s]扩容后: 当前副本数:%d\n", app_name, replicas)
					}
				}
			}
		}
	}

}

func fallback_app_replicas(v *viper.Viper, app_name string) {
	/**
		@brief 回退应用副本数
		@param app string 应用名称
	**/
	username := v.GetStringMapString("database.ssh")["username"]
	password := v.GetStringMapString("database.ssh")["password"]
	port_str := v.GetStringMapString("database.ssh")["port"]
	port, _ := strconv.Atoi(port_str)
	replicas := v.Get("database.replicas").(map[string]interface{})
	// 检查 appName 是否存在于 replicas 中
	if service, exists := replicas[app_name]; exists {
		// 将 service 转换为切片
		for _, serviceConfig := range service.([]interface{}) {
			// if serviceConfig.(map[string]interface{}).["enable"] {
			config := serviceConfig.(map[string]interface{})
			if config["enable"].(bool) {
				ip := config["ip"].(string)
				min := config["min"].(int)
				// max := config["max"]
				replicas, _ := get_app_replicas(ip, username, password, port, app_name)
				fmt.Printf("[%s]缩容前: 当前副本数:%d\n", app_name, replicas)
				if replicas > min {
					set_app_replicas(ip, username, password, port, app_name, replicas-1)
					app_restart(ip, username, password, port, app_name)
					replicas, _ = get_app_replicas(ip, username, password, port, app_name)
					fmt.Printf("[%s]扩容后: 当前副本数:%d\n", app_name, replicas)
				}
			}
		}
	}
}

func main() {
	current_dir, err := os.Getwd()
	if err != nil {
		fmt.Println("获取脚本当前目录失败")
		return
	}
	config_file_name := "config/config.yaml"
	config_file_path := filepath.Join(current_dir, config_file_name)
	v := viper.New()
	v.AddConfigPath(current_dir)
	v.SetConfigFile(config_file_path)
	v.SetConfigType("yaml")
	if err := v.ReadInConfig(); err != nil {
		fmt.Println("配置文件读取错误:", err)
		return
	}
	kafka_brokers := v.GetStringMapString("database.kafka")["brokers"]
	lag_str := v.GetStringMapString("database.kafka")["lag"]
	lag, _ := strconv.ParseInt(lag_str, 10, 64)

	for {
		groups, _ := get_kafka_lag(strings.Split(kafka_brokers, ","), lag)
		for groupId := range groups {
			switch groupId {
			case "cleaners_groupid":
				deploy_app_replicas(v, "cleaner")
			case "groupid_dataservice":
				deploy_app_replicas(v, "transfer")
			case "event_data_preparation":
				deploy_app_replicas(v, "security-event")
			case "groupid_threat":
				deploy_app_replicas(v, "threat")
			case "groupid_running_distribution":
				deploy_app_replicas(v, "analyzer-dev")
			default:
				fmt.Println("未知主题, 无需扩容")
			}
		}
		wxgz_groups := []string{"cleaner", "transfer", "security-event", "threat", "analyzer-dev"}
		if len(groups) < len(wxgz_groups) {
			fallback_app_tag := []string{}
			for groupId := range groups {
				switch groupId {
				case "cleaners_groupid":
					fallback_app_tag = append(fallback_app_tag, "cleaner")
				case "groupid_dataservice":
					fallback_app_tag = append(fallback_app_tag, "transfer")
				case "event_data_preparation":
					fallback_app_tag = append(fallback_app_tag, "security-event")
				case "groupid_threat":
					fallback_app_tag = append(fallback_app_tag, "threat")
				case "groupid_running_distribution":
					fallback_app_tag = append(fallback_app_tag, "analyzer-dev")
				default:
					fmt.Println("未知主题, 无需缩容")
				}
			}
			for _, app := range wxgz_groups {
				// 如果_app不存在于fallback_app_tags中，执行缩容操作
				if !slices.Contains(fallback_app_tag, app) {
					fallback_app_replicas(v, app)
					time.Sleep(10 * time.Second)
				}
			}
		}
		time.Sleep(5 * time.Minute)
	}
}

File: python/ssh/bin/auto-replicas.py
import os,yaml
import paramiko
from io import StringIO
import time
from concurrent.futures import ThreadPoolExecutor
from concurrent.futures import ThreadPoolExecutor
from kafka import KafkaConsumer, TopicPartition
from kafka import KafkaConsumer, TopicPartition
from kafka import KafkaAdminClient
from kafka.structs import TopicPartition


def get_consumer_lag(kafka_url,topic,group_id,lag=100000,consecutive_threshold=3):
    """
    获取 Kafka 消费者组的滞后信息。

    Args:
        kafka_url (str): Kafka 服务器地址。
        topic (str): 要监控的 Kafka 主题。
        group_id (str): Kafka 消费者组 ID。
        lag (int): 滞后阈值。
        consecutive_threshold (int): 连续滞后次数阈值，需要连续大于 lag 的次数才能返回。

    Returns:
        str: 消费者组 ID，如果连续滞后次数大于等于阈值，否则返回 None。
    """
    # 配置 Kafka 消费者
    consumer = KafkaConsumer(
        # 'app_information',  # 替换为您要监控的 Kafka 主题
        group_id=group_id,  # 替换为您的 Kafka 消费者组
        bootstrap_servers=kafka_url,  # 替换为您的 Kafka 服务器地址
    )
    # 获取分区列表
    partitions = consumer.partitions_for_topic(topic)
    if partitions is None:
        raise ValueError("Topic not found")
    
    consecutive_lags = 0
    for _ in range(consecutive_threshold):
        # 初始化lag为0
        lags = 0
        # 遍历每个分区
        for partition in partitions:
            tp = TopicPartition(topic, partition)

            # 分配分区并获取分区结束位置
            consumer.assign([tp])
            consumer.seek_to_end(tp)
            end_offset = consumer.position(tp)

            # 获取分区的 CURRENT-OFFSET（消费者当前位置）
            consumer.seek_to_beginning(tp)
            current_offset = consumer.position(tp)

            # 计算 LAG（滞后）
            kafka_lag = end_offset - current_offset
            lags += kafka_lag
        if lags >= lag:
            consecutive_lags += 1
            time.sleep(2)
        else:
            consecutive_lags = 0
    if consecutive_lags >= consecutive_threshold:
        # print(f"Topic: {topic}, Partition: {partition}, CURRENT-OFFSET: {current_offset}, LOG-END-OFFSET: {end_offset}, LAG: {lag}")
        # return(f"GroupId: {group_id},Topic: {topic}, LAGS总数: {lags}")
        return(group_id)
    
        # return(group_id,topic,lags)
    # 关闭 Kafka 消费者
    consumer.close()

def get_kafka_msg(kafka_url,lag):
    """
    获取 Kafka 消费者组的滞后信息
    
    Args:
        kafka_url (str): Kafka 服务器地址
        lag (int): 滞后阈值
    
    Returns:
        list: 消费者组 ID，如果连续滞后次数大于等于阈值，否则返回 None
    """
    # 配置 Kafka 集群的连接信息
    try:
        admin_client = KafkaAdminClient(
            bootstrap_servers=kafka_url,  # 替换为您的 Kafka 服务器地址
        )
    except Exception as e:
        return [f"kafka连接失败:{e}"]
    # 获取 Kafka 集群中的所有主题
    topics = admin_client.list_topics()
    # print("所有主题:")
    # for topic in topics:
    #     print(topic)


    # 获取消费者组信息
    consumer_groups = admin_client.list_consumer_groups()
    wxgz_groups = ["groupid_threat","groupid_dataservice","event_data_preparation","cleaners_groupid","groupid_running_distribution"]
    # print("所有消费者组:")
    # for consumer_group in consumer_groups:
    #     print(consumer_group[0])

    # # 获取每个消费者组订阅的主题
    # print("消费者组和它们订阅的主题:")
    # for consumer_group in consumer_groups:
    #     group_description = admin_client.list_consumer_groups(consumer_group[0])
    #     # 打印消费者组和其订阅的主题
    #     print(f"Consumer Group: {consumer_group[0]}, Subscribed Topics: {group_description}")
    #     # for i in group_description.topics:
    #         # print(f"Consumer Group: {consumer_group[0]}, Subscribed Topics: {i}")

    unique_topics = set()
    result_seds = []
    # 遍历每个消费者组
    for group in consumer_groups:
        if group[0] in wxgz_groups:
            # 获取消费者组订阅的主题
            topics = admin_client.list_consumer_group_offsets(group[0])
            # 处理返回的 TopicPartition 列表
        
            for tp in topics:
                # topic_name = tp.topic
                # print(f"Consumer Group: {group[0]}, Subscribed Topic: {topic_name}")
                topic_name = tp.topic
                unique_topics.add(topic_name)
            # 打印唯一的主题名称
            # for topic_name in unique_topics:
            #     # print(f"Consumer Group: {group[0]}, Subscribed Topic: {topic_name}")
            #     get_consumer_lag(kafka_url,topic_name,group[0],limit)
            # 使用 ThreadPoolExecutor 来并发验证
            with ThreadPoolExecutor(max_workers=len(unique_topics)) as executor:
            # with ThreadPoolExecutor(max_workers=999) as executor:
                futures = [executor.submit(get_consumer_lag, kafka_url,topic_name,group[0],lag) for topic_name in unique_topics]
                # 收集结果
                results = [future.result() for future in futures]

            # 打印验证结果
            for result in results:
                if result is not None:
                    result_seds.append(result)
                    # print(result)
    return result_seds

def get_remote_system_info(ip, username, password, port):
    """
    获取远程系统信息

    Args:
        ip (str): IP 地址
        username (str): 用户名
        password (str): 密码
        port (int): 端口

    Returns:
        tuple: CPU 信息，内存信息，磁盘信息
    """
    # 创建ssh客户端
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    try:
        # 连接远程服务器
        ssh_client.connect(hostname=ip, username=username, password=password, port=port)
        
        # 执行命令
        stdin, stdout, stderr = ssh_client.exec_command("top -b -n 1 | grep 'Cpu(s)'")
        cpuinfo = stdout.read().decode("utf-8").split(",")[3].split()[0]
        # print("cpu信息:" + cpuinfo)

        stdin, stdout, stderr = ssh_client.exec_command("free -g|grep Mem")
        meminfo = stdout.read().decode("utf-8").split()[-1]
        # print("meminfo: " + meminfo)

        stdin, stdout, stderr = ssh_client.exec_command("df -h -BG /home/app | sort -k 4 -h|tail -n 1")
        disk_output = stdout.read().decode("utf-8").split()[3].rstrip('G')
        # print("disk_output: " + disk_output)
        return cpuinfo,meminfo,disk_output

    finally:
        # 关闭连接
        ssh_client.close()

def get_app_replicas(ip, username, password, port,app_name):
    """
    获取指定应用的副本数
    :param ip: 远程服务器IP
    :param username: 用户名
    :param password: 密码
    :param port: 端口
    :param app_name: 应用名称
    :return: 应用副本数
    """
    # 创建ssh客户端
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    try:
        # 连接远程服务器
        ssh_client.connect(hostname=ip, username=username, password=password, port=port)

        # 使用sftp传输文件
        sftp= ssh_client.open_sftp()
        with sftp.file("/home/app/app/"+app_name+"/bin/docker-compose.yml","r") as remote_file:
            compose_data = yaml.safe_load(remote_file)
        # 更新指定服务的环境变量
        for service in compose_data["services"].values():
            try:
                # print(service["deploy"]["replicas"])
                return(service["deploy"]["replicas"])
                break
            except:
                pass
    finally:
        sftp.close()
        ssh_client.close()

def set_app_replicas(ip, username, password, port, app_name,replicas):
    """
    设置指定应用的副本数

    Args:
        ip (str): IP 地址
        username (str): 用户名
        password (str): 密码
        port (int): 端口
        app_name (str): 应用名称
        replicas (int): 副本数

    Returns:
        None: 无返回值
    """
    # 创建ssh客户端
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    try:
        # 连接远程服务器
        ssh_client.connect(hostname=ip, username=username, password=password, port=port)

        # 使用sftp传输文件
        sftp= ssh_client.open_sftp()
        with sftp.file("/home/app/app/"+app_name+"/bin/docker-compose.yml","r") as remote_file:
            compose_data = yaml.safe_load(remote_file)
        # 更新指定服务的环境变量
        for service in compose_data["services"].values():
            try:
                service["deploy"]["replicas"] = replicas
                break
            except:
                pass
        # 将更新后的内容写回到远程文件
        updated_data = StringIO()
        yaml.dump(compose_data, updated_data, default_flow_style=False)
        updated_data.seek(0)
        with sftp.file("/home/app/app/"+app_name+"/bin/docker-compose.yml", 'w') as remote_file:
            remote_file.write(updated_data.read())
    finally:
        # 关闭连接
        sftp.close()
        ssh_client.close()

def deploy_app_replicas(app):
    """
    部署应用副本数
    Args:
        app (str): 应用名称

    Returns:
        None: 无返回值
    """
    replicas_list = config["database"]["replicas"]
    username = config["database"]["ssh"]["username"]
    password = config["database"]["ssh"]["password"]
    port = config["database"]["ssh"]["port"]
    for service in config["database"]["replicas"][app]:
        if service["enable"]:
            ip = service["ip"]
            max = service["max"]
            min = service["min"]
            cpuinfo,meminfo,disk_output = get_remote_system_info(ip, username,password,port)
            if float(cpuinfo) >= 20 and int(meminfo) >= 2 and int(disk_output) >= 20:
                print(f"{ip} cpuinfo: {cpuinfo} meminfo: {meminfo} disk_output: {disk_output}")
                replicas = get_app_replicas(ip, username,password,port,app)
                print(f"扩容:{app}当前副本数:{str(replicas)}")
                if int(replicas) < int(max):
                    set_app_replicas(ip, username,password,port,app,int(replicas)+1)
                    app_restart(ip, username,password,port,app)
                replicas = get_app_replicas(ip, username,password,port,app)
                print(f"扩容:{app}当前副本数:{str(replicas)}")

def fallback_app_replicas(app):
    """
    回滚应用副本数

    Args:
        app (str): 应用名称

    Returns:
        None: 无返回值
    """
    replicas_list = config["database"]["replicas"]
    username = config["database"]["ssh"]["username"]
    password = config["database"]["ssh"]["password"]
    port = config["database"]["ssh"]["port"]
    for service in config["database"]["replicas"][app]:
        if service["enable"]:
            ip = service["ip"]
            max = service["max"]
            min = service["min"]
            replicas = get_app_replicas(ip, username,password,port,app)
            print(f"缩容:{app}当前副本数:{str(replicas)}")
            if int(replicas) > int(min):
                set_app_replicas(ip, username,password,port,app,int(replicas)-1)
                app_restart(ip, username,password,port,app)
            replicas = get_app_replicas(ip, username,password,port,app)
            print(f"缩容:{app}当前副本数:{str(replicas)}")

def app_restart(ip, username, password, port, app_name):
    """
    应用重启

    Args:
        ip (str): IP 地址
        username (str): 用户名
        password (str): 密码
        port (int): 端口
        app_name (str): 应用名称

    Returns:
        None: 无返回值
    """
    # 创建ssh客户端
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    try:
        # 连接远程服务器
        ssh_client.connect(hostname=ip, username=username, password=password, port=port)
        
        # 执行命令
        stdin, stdout, stderr = ssh_client.exec_command("/home/app/bin/docker-compose -f /home/app/app/"+app_name+"/bin/docker-compose.yml down ")
        stdout.channel.recv_exit_status()  # 等待命令执行完成
        # time.sleep(5)
        # print("停止信息:" + stdout)
        stdin, stdout, stderr = ssh_client.exec_command("/home/app/bin/docker-compose -f /home/app/app/"+app_name+"/bin/docker-compose.yml up -d")
        stdout.channel.recv_exit_status()  # 等待命令执行完成
        # print("启动信息:" + stdout)

    finally:
        ssh_client.close()

def main(config):
    replicas_tag = 0
    # replicas_list = config["database"]["replicas"]
    # username = config["database"]["ssh"]["username"]
    # password = config["database"]["ssh"]["password"]
    # port = config["database"]["ssh"]["port"]
    # for service_list in replicas_list:
    #     for service in replicas_list.get(service_list):
    #         if service["enable"]:
    #             ip = service["ip"]
    #             cpuinfo,meminfo,disk_output = get_remote_system_info(ip, username,password,port)
    #             print(f"{ip} service_name: {service_list}  cpuinfo: {cpuinfo} meminfo: {meminfo} disk_output: {disk_output}")

    # for service_list in replicas_list:
    #     for service in replicas_list.get(service_list):
    #         if service["enable"]:
    #             ip = service["ip"]
    #             min = service["min"]
    #             max = service["max"]
    #             print(f"{ip} service_name: {service_list} min: {min} max: {max}")
    while True:
        kafka_msg = get_kafka_msg(config["database"]["kafka"]["brokers"], config["database"]["kafka"]["lag"])
        if len(kafka_msg) != 0:
            replicas_tag = 0
            unique_topics = set()
            for  index, value in enumerate(kafka_msg):
                unique_topics.add(value)
            for i in unique_topics:
                match i:
                    case "cleaners_groupid":
                        deploy_app_replicas("cleaner")
                    case "groupid_dataservice":
                        deploy_app_replicas("transfer")
                    case "event_data_preparation":
                        deploy_app_replicas("security-event")
                    case "groupid_threat":
                        deploy_app_replicas("threat")
                    case "groupid_running_distribution":
                        deploy_app_replicas("analyzer-dev")
                    case _:
                        print("未知主题,无需扩容")
        else:
            replicas_tag += 1
            if replicas_tag >= 3:
                wxgz_groups = ["cleaner", "transfer", "security-event", "threat", "analyzer-dev"]
                unique_topics = set()
                for  index, value in enumerate(kafka_msg):
                    unique_topics.add(value)
                fallback_app_tag = []
                for i in unique_topics:
                    match i:
                        case "cleaners_groupid":
                            fallback_app_tag.append("cleaner")
                        case "groupid_dataservice":
                            fallback_app_tag.append("transfer")
                        case "event_data_preparation":
                            fallback_app_tag.append("security-event")
                        case "groupid_threat":
                            fallback_app_tag.append("threat")
                        case "groupid_running_distribution":
                            fallback_app_tag.append("analyzer-dev")
                        case _:
                            print("未知主题,无需扩容")
                for app in wxgz_groups:
                    # 如果 app 不在 fallback_app_tag 中，则执行缩容操作
                    if app not in fallback_app_tag:
                        fallback_app_replicas(app)
                        time.sleep(10)  # 休眠 10 秒
                replicas_tag = 0
        time.sleep(10)

if __name__ == "__main__":
    config_file_path = "运维/ssh/config/config.yaml"
    abs_path = os.path.abspath(config_file_path)
    print(abs_path)

    # config_data = read_config_file(config_file_path)
    with open(config_file_path, "r") as yamlfile:
        config = yaml.safe_load(yamlfile)
    main(config)
